File: 024_SwapNodesInPairs.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
//6.45%
struct ListNode* swapPairs(struct ListNode* head) {
    //Check boundary conditions
	if(!head || !head -> next)
		return head;
	
	struct ListNode *prev = head, *curr = head -> next;
	prev -> next = curr -> next;
	curr -> next = prev;
	head = curr;
	//prev = prev -> next;

	while(prev -> next){
		curr = prev -> next;
		if(curr -> next){
			struct ListNode *temp = curr -> next -> next;
			curr -> next -> next = curr;
			prev -> next = curr -> next;
			curr -> next = temp;
			prev = curr;
		}
		else
			break;
	}
	return head;
}

//6.45%
//Recursive way
struct ListNode* swapPairs(struct ListNode* head) {
    //Check boundary conditions
	if(!head || !head -> next)
		return head;
	
	struct ListNode *prev = head, *curr = head -> next;
	prev -> next = swapPairs(curr -> next);
	curr -> next = prev;
	return curr;	
}

File: 024_SwapNodesInPairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# From LeetCode solutions
def swapPairs(self, head):
    pre, pre.next = self, head
    while pre.next and pre.next.next:
        a = pre.next
        b = a.next
        pre.next, b.next, a.next = b, a, b.next
        pre = a
    return self.next
	
# 52.5%
class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head
        curr = head.next
        head.next, curr.next = self.swapPairs(curr.next), head
        return curr